core-api/src/main/java/org/silverpeas/core/cache/service/ThreadCache.java

```diff
@@ -34,7 +34,7 @@
  */
 class ThreadCache extends AbstractSimpleCache {
 
-  private final ThreadLocal<Map<Object, Object>> cache = new ThreadLocal<>();
+  private static final ThreadLocal<Map<Object, Object>> cache = new ThreadLocal<>();
```

Comment:
I agree about that it is a recommandation when the cache is linked to a unique functional or technical stuff. But here, as it has been chosen to handle one cache linked to requests and an other one to handledcache linked to batch processes. So, it introduces breaking change into technical part of Silverpeas. Otherwise CacheServiceProvider.getThreadCacheService() or CacheServiceProvider.getRequestCacheService() have to disapear on order to use only one of them.

Reply:
In fact there is not breaking changes. `ThreadLocal` and `ThreadCache` are two different kinds. Explanation:

- `TheadCacheService` uses a `TheadCache` to manage a cache for each thread in Silverpeas.
- `RequestCacheService` uses a `ThreadCache` to manage a cache for each incoming request because in JEE each incoming request is processed by a different thread. But, for a caller, it is an implementation detail the `RequestCacheService` uses a `ThreadCache` as cache.
- `ThreadCache` provides an interface for caller to cache data into the cache of the current thread. For doing `ThreadCache` uses a `ThreadLocal` field.
- With `ThreadLocal`, you can have several instances of `ThreadCache` or whatever objects with a `ThreadLocal` field. Whatever it is done, `ThreadLocal` will refer always the current thread. But, because there is OS interactions by the JVM when storing data into the thread caches, it seams there are side effects when `ThreadLocal` is an instance field instead of a static field. Thread cache handling in ThreadLocal implies hardware interactions, hence C-binding with the OS.  The side effects were visible in the test because a double call to the cache cleaning method was required to ensure the thread cache to be correctly cleaned.

Reply:
After a physical and verbal conversation, we finally understood each other.